File: 1790_Check_if_One_String_Swap_Can_Make_Strings_Equal.py
class Solution:
    def areAlmostEqual(self, s1: str, s2: str) -> bool:
        if s1 == s2:
            return True

        n = len(s1)

        stk = []
        for i in range(n):
            if s1[i] != s2[i]:
                stk.append(i)
            if len(stk) > 2:
                return False
        
        if len(stk) == 2:
            if s1[stk[0]] == s2[stk[1]] and s1[stk[1]] == s2[stk[0]]:
                return True

        return False

File: 1790_Check_if_One_String_Swap_Can_Make_Strings_Equal.java
class Solution {
    public boolean areAlmostEqual(String s1, String s2) {
        if (s1.equals(s2)) {
            return true;
        }

        int n = s1.length();
        Stack<Integer> stk = new Stack<>();
        for (int i = 0; i < n; i++) {
            if (s1.charAt(i) != s2.charAt(i)) {
                stk.push(i);
            }
            if (stk.size() > 2) {
                return false;
            }
        }
        if (stk.size() == 2) {
            int first = stk.pop();
            int second = stk.pop();
            return s1.charAt(first) == s2.charAt(second) && s1.charAt(second) == s2.charAt(first);
        }

        return false;

    }
}
